Fix additional negative time issues in ROS Kilted

JayHerpin opened the pull request:
Fixes a few more cases that https://github.com/cra-ros-pkg/robot_localization/pull/943 missed

Commit:
Fix additional negative time issues in ROS Kilted

The previous fix (e992754) addressed some Time constructor calls but
missed additional code paths:

1. getFilteredAccelMessage(): Removed redundant rclcpp::Time() wrapper
   around filter_.getLastMeasurementTime() which already returns an
   rclcpp::Time with RCL_ROS_TIME clock type. Now matches the pattern
   used in getFilteredOdometryMessage().

2. preparePose(): Added RCL_ROS_TIME clock type to Time constructor
   calls for lookupTransformSafe(), matching the pattern used in
   navsat_transform.cpp from commit e992754.

These paths are only exercised with specific configurations (e.g.,
acceleration output enabled, certain frame transform setups), which
is why they weren't caught in the original fix.

Commit:
Fix negative time crash in periodicUpdate clearExpiredHistory

Guard against subtracting history_length_ from last_measurement_time
when last_measurement_time is smaller, which results in negative time
and throws an exception in rclcpp::Time.

This occurs during simulation startup when the filter has received
measurements but hasn't been running long enough for history_length_
(default 0.5s) to have elapsed.

SteveMacenski commented on src/ros_filter.cpp:
```diff
@@ -2965,7 +2968,7 @@ bool RosFilter<T>::preparePose(
   bool can_transform = ros_filter_utilities::lookupTransformSafe(
     tf_buffer_.get(), final_target_frame, pose_tmp.frame_id_,
-    rclcpp::Time(tf2::timeToSec(pose_tmp.stamp_)), tf_timeout_,
+    rclcpp::Time(tf2::timeToSec(pose_tmp.stamp_), RCL_ROS_TIME), tf_timeout_,
```
Shouldn't we use the `node->get_clock()` type of time? This assumes simulation time.

JayHerpin replied:
Its my understanding that when not run in simulation mode RCL_ROS_TIME reports the same as RCL_SYSTEM_TIME. Is that incorrect? At least some of the choices made here were matching the conventions from this commit https://github.com/cra-ros-pkg/robot_localization/commit/c20b934daff5269e32dfb6ad646af1423b8e01f4 which seems to be the first pass at fixing some of these issues

ayrton04 replied:
https://design.ros2.org/articles/clock_and_time.html

> The ROSTime will report the same as SystemTime when a ROS Time Source is not active. When the ROS time source is active ROSTime will return the latest value reported by the Time Source. ROSTime is considered active when the parameter use_sim_time is set on the node.

So this is correct, and is the same mechanism as ROS 1, I believe. What confuses me is why the `rclcpp::Time` [API](https://api.nav2.org/jazzy/html/classrclcpp_1_1Time.html) defaults to `SYSTEM_TIME` for any constructor except the one that takes a `Time` message. It also seems to violate the design documentation:

> It is expected that the default choice of time will be to use the ROSTime source, however the parallel implementations supporting steady_clock and system_clock will be maintained for use cases where the alternate time source is required.

SteveMacenski commented on src/ros_filter.cpp:
```diff
@@ -2320,7 +2320,10 @@ void RosFilter<T>::periodicUpdate()
   if (smooth_lagged_data_) {
-    clearExpiredHistory(filter_.getLastMeasurementTime() - history_length_);
+    const rclcpp::Time last_measurement_time = filter_.getLastMeasurementTime();
+    if (last_measurement_time.nanoseconds() > history_length_.nanoseconds()) {
```
I think you can just do a direct `rclcpp::Time` comparison https://docs.ros2.org/foxy/api/rclcpp/classrclcpp_1_1Time.html

JayHerpin replied:
history_length_ is actually a rclcpp::Duration, which is not directly comparable with rclcpp::Time... but given your response, clearly the code needs a clarifying comment about why this check exists at all given you wouldn't normally compare a duration and time

SteveMacenski commented on src/ros_filter.cpp:
```diff
@@ -532,7 +532,7 @@ bool RosFilter<T>::getFilteredAccelMessage(
 
     // Fill header information
-    message->header.stamp = rclcpp::Time(filter_.getLastMeasurementTime());
+    message->header.stamp = filter_.getLastMeasurementTime();
```
Any functional change here or just clarity? I don't see any changes since that cast will happen implicitly, but wanted to make sure there isn't something subtle here we should be aware of

Commit:
Address PR review feedback

- Revert unintentional removal of rclcpp::Time cast in getFilteredAccelMessage
- Add clarifying comment for negative time point prevention check

JayHerpin commented:
I'm not clear if there is a requested change for me here or not. Is there anything I can do to move this along?

ayrton04 commented:
No, sorry, just never merged it.

I need to revisit the time management in this package. When I developed the workshop for ROSCon UK last year, I found that it didn't work well with bags/sim. I [changed the node's clock type](https://github.com/cra-ros-pkg/robot_localization/pull/943/changes), but I think I read somewhere that this should not be done unless you know what you're doing. This leads me to think I should have gone about it in a different way.

ayrton04 commented:
Would you mind PR-ing into rolling as well? Thanks!
